core/src/main/java/dk/alexandra/fresco/framework/network/socket/Connector.java

```diff
@@ -0,0 +1,175 @@
+package dk.alexandra.fresco.framework.network.socket;
+
+import dk.alexandra.fresco.framework.Party;
+import dk.alexandra.fresco.framework.configuration.NetworkConfiguration;
+import java.io.IOException;
+import java.net.ConnectException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.net.ServerSocketFactory;
+import javax.net.SocketFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Connector implements NetworkConnector {
+
+  public static final Duration DEFAULT_CONNECTION_TIMEOUT = Duration.ofMinutes(1);
+  private static final int PARTY_ID_BYTES = 1;
+  private static final Logger logger = LoggerFactory.getLogger(SocketNetwork.class);
+  private final Map<Integer, Socket> socketMap;
+  private final SocketFactory socketFactory;
+  private final ServerSocketFactory serverFactory;
+
+  Connector(NetworkConfiguration conf, Duration timeout) {
+    this(conf, timeout, SocketFactory.getDefault(), ServerSocketFactory.getDefault());
+  }
+
+  Connector(NetworkConfiguration conf, Duration timeout,
+      SocketFactory socketFactory, ServerSocketFactory serverFactory) {
+    this.socketFactory = socketFactory;
+    this.serverFactory = serverFactory;
+    this.socketMap = connectNetwork(conf, timeout);
+  }
+
+  @Override
+  public Map<Integer, Socket> getSocketMap() {
+    return this.socketMap;
+  }
+
+  /**
+   * Fully connects the network.
+   * <p>
+   * Connects a channels to each external party (i.e., parties other than this party).
+   * </p>
+   *
+   * @param conf the configuration defining the network to connect
+   * @param timeout duration to wait until timeout
+   * @return a map from party ids to the associated communication channel
+   */
+  private Map<Integer, Socket> connectNetwork(final NetworkConfiguration conf,
+      final Duration timeout) {
+    Map<Integer, Socket> socketMap = new HashMap<>(conf.noOfParties());
+    ExecutorService es = Executors.newFixedThreadPool(2);
```

Comment:
Wouldn't that risk the code "hanging" if the one of the clients fail somehow?

Reply:
No, the code would be the same as now, but two pools rather than one with two threads.

Reply:
What would be a reason for having the server not timeout? I think I could make it work, I am just not sure I understand the reasoning. 

Also, what is the benefit of having two separate pools (or `ExecutorServices`s I guess)? It seems to me that this would require more work to get the same functionality. One of the things I am trying to do here is to make sure that we fail as early as possible. I.e., if the server fails we cancel the clients immediately instead of waiting for the clients to timeout, and the other way around (since if either client or server connections fail we are not able to connect the Network). That is what I am using the `CompletionService` for. However, I cant use the `CompletionService` with two different `Executors`. So it seems to me I would end up having to roll my own `CompletionService` variant.